### PR TITLE
d3q27 BGK with Kuperstokh

### PR DESCRIPTION
**Pull request by wojtasMEiL**
I added the model D3Q27 BGK (the standard one) with Kuperstokh forcing term
Kuperstokh forcing term added in an "if" statement. No thread divergence expected due to global force implementation.
Also, two 3d_channel_test files added (unsure if the first one existed before)

### CODE REVIEW

File: src/d3q27_BGK/Dynamics.c [hunk 1]
```diff
@@ -0,0 +1,545 @@
+/*-------------------------------------------------------------*/
+/*  CLB - Cudne LB - Stencil Version                           */
+/*     CUDA based Adjoint Lattice Boltzmann Solver             */
+/*     Author: Lukasz Laniewski-Wollk                          */
+/*     Developed at: Warsaw University of Technology - 2012    */
+/*-------------------------------------------------------------*/
+
```
Comment (mdzik) [inline src/d3q27_BGK/Dynamics.c]: Missing developer info
Missing original reference
see https://github.com/CFD-GO/TCLB/blob/develop/.github/CONTRIBUTING.md

**Comment by mdzik** [comment]
missing references. 
see https://github.com/CFD-GO/TCLB/blob/develop/.github/CONTRIBUTING.md

File: src/d3q27_BGK_gallcor/Dynamics.c [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+		delta f_force = f_eq(u+du) - f_eq(u) where du = u+force/rho
+
+		"Pressure drop in flow across ceramic foams--A numerical and experimental study"
+		
+		The original idea comes from:
+		
+		Kupershtokh, A., Medvedev, D., Karpov, D., 2009. "On equations of state in a lattice Boltzmann method". Comput. Math. Appl. 58 (5), 965-974
+
+		*/
+
```
Comment (mdzik) [inline src/d3q27_BGK_gallcor/Dynamics.c]: Sorry.
Move/copy that to the top

File: src/d3q27_BGK_gallcor/Dynamics.c [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+        f112 =(1.-omega)*f112 + omega*( ( rho + ( -Jz + Jy + Jx + ( Jz*Jz + Jy*Jy + Jx*Jx + ( -Jz*Jy + ( -Jz + Jy )*Jx )*3. )/rho )*3. )*0.00462962962962963);
+        f212 =(1.-omega)*f212 + omega*( ( rho + ( -Jz + Jy - Jx + ( Jz*Jz + Jy*Jy + Jx*Jx + ( -Jz*Jy + ( Jz - Jy )*Jx )*3. )/rho )*3. )*0.00462962962962963);
+        f022 =(1.-omega)*f022 + omega*( rho*0.0185185185185185 + ( -Jx*Jx/rho + ( -Jz - Jy + ( Jz*Jz + ( Jy + Jz*3. )*Jy )/rho )*2. )/36.);
+        f122 =(1.-omega)*f122 + omega*( ( rho + ( -Jz - Jy + Jx + ( Jz*Jz + Jy*Jy + Jx*Jx + ( Jz*Jy + ( -Jz - Jy )*Jx )*3. )/rho )*3. )*0.00462962962962963);
+        f222 =(1.-omega)*f222 + omega*( ( rho + ( -Jz - Jy - Jx + ( Jz*Jz + Jy*Jy + Jx*Jx + ( Jz*Jy + ( Jz + Jy )*Jx )*3. )/rho )*3. )*0.00462962962962963);
+	
+	
+	/* this conditional statement is either performed on the whole GPU or not because forces are defined globally */
+	if( ForceX*ForceX + ForceY*ForceY + ForceZ*ForceZ > 1e-16 ){
+
```
Comment (mdzik) [inline src/d3q27_BGK_gallcor/Dynamics.c]: Floating point caparison. Maybe boolean/byte flag?
Force could be zonal

**Comment by wojtasMEiL** [comment]
Applied necessary corrections: references + author reference. Also, the conditional force term slightly modified, ŁŁW says it will be ok. No sense adding another flag in such a simple model.

File: src/d3q27_BGK_gallcor/Dynamics.R [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+AddQuantity( name="P",unit="Pa")
+AddQuantity( name="U",unit="m/s",vector=T)
+
+AddSetting(name="nu", default=0.16666666, comment='Viscosity')
+AddSetting(name="Velocity", default="0m/s", comment='Inlet velocity', zonal=TRUE)
+AddSetting(name="Pressure", default="0Pa", comment='Inlet pressure', zonal=TRUE)
+
+AddSetting(name="GalileanCorrection",default=0.,comment='Galilean correction term')
```
Comment (mdzik) [inline src/d3q27_BGK_gallcor/Dynamics.R]: This flag does not seem to be used

File: src/d3q27_BGK_gallcor/Dynamics.R [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	dy   = U[,2],
+	dz   = U[,3],
+	comment=paste("density F",1:27-1),
+	group="f"
+)
+
+AddQuantity( name="P",unit="Pa")
+AddQuantity( name="U",unit="m/s",vector=T)
+
+AddSetting(name="nu", default=0.16666666, comment='Viscosity')
```
Comment (mdzik) [inline src/d3q27_BGK_gallcor/Dynamics.R]: Missing unit

File: src/d3q27_BGK_gallcor/Dynamics.R [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+AddSetting(name="Velocity", default="0m/s", comment='Inlet velocity', zonal=TRUE)
+AddSetting(name="Pressure", default="0Pa", comment='Inlet pressure', zonal=TRUE)
+
+AddSetting(name="GalileanCorrection",default=0.,comment='Galilean correction term')
+AddSetting(name="ForceX",default=0, comment='Force X')
+AddSetting(name="ForceY",default=0,comment='Force Y')
+AddSetting(name="ForceZ",default=0,comment='Force Z')
+
```
Comment (mdzik) [inline src/d3q27_BGK_gallcor/Dynamics.R]: Missing units?

**Comment by mdzik** [comment]
Add same/more (?) refs to src/d3q27_BGK_gallcor. 
Do we need two (not so) different models? 
It is possible to move only some files to branch and then make pull request

**Comment by wojtasMEiL** [comment]
It is ready. You may merge it.

**Pull request merged**